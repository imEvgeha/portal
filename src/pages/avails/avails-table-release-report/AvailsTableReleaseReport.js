import React, {useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import Button from '@atlaskit/button';
import Popup from '@atlaskit/popup';
import Select from '@atlaskit/select';
import AtlaskitMoreIcon from '@vubiquity-nexus/portal-assets/atlaskit-more-icon.svg';
import withToasts from '@vubiquity-nexus/portal-ui/lib/toast/hoc/withToasts';
import {downloadFile} from '@vubiquity-nexus/portal-utils/lib/Common';
import moment from 'moment';
import {exportService} from '../../legacy/containers/avail/service/ExportService';
import {
    PRE_PLAN_SELECTED_TAB,
    PRE_PLAN_TAB,
    RIGHTS_SELECTED_TAB,
    RIGHTS_TAB,
    SELECTED_FOR_PLANNING_TAB,
    STATUS_TAB
} from '../rights-repository/constants';
import {CREATE_REPORT, ERROR_MESSAGE, MOCK_YEAR, MONTHS, START_YEAR, END_YEAR, NEW_RELEASE_REPORT} from './constants';
import './AvailsTableReleaseReport.scss';

const AvailsTableReleaseReport = ({addToast, activeTab, selectedRows, totalRows, prePlanRightsCount, planningRightsCount}) => {
    const [isLoading, setIsLoading] = useState(false);
    const [isOpen, setIsOpen] = useState(false);
    const [isDisabled, setIsDisabled] = useState(false);
    const [selectedMonth, setSelectedMonth] = useState(moment().format('MMMM'));
    const [selectedYear, setSelectedYear] = useState(moment().format('YYYY'));

    useEffect(() => {
        let disable = false;
        const isItEmptyRightsTab = [RIGHTS_TAB, RIGHTS_SELECTED_TAB].includes(activeTab) && totalRows === 0;
        const isItDisabledForCurrentTab = [PRE_PLAN_TAB, PRE_PLAN_SELECTED_TAB, SELECTED_FOR_PLANNING_TAB, STATUS_TAB].includes(activeTab);
        
        if (isItEmptyRightsTab || isItDisabledForCurrentTab) {
            disable = true
        } 
        
        setIsDisabled(disable);
    }, [activeTab, selectedRows, totalRows, prePlanRightsCount]);

    const onCreateReport = () => {
        const monthNumber = moment().month(selectedMonth).format('MM');
        const yearMonth = `${selectedYear}-${monthNumber}`;
        const startFrom = `${yearMonth}-01T00:00:00.000`;
        const endDay = moment(yearMonth, 'YYYY-MM').daysInMonth();
        const startTo = `${yearMonth}-${endDay}T23:59:59.999`;
        setIsLoading(true);
        exportService
            .getReleaseReport({startFrom, startTo})
            .then(response => {
                downloadFile(response, `${selectedMonth}-${selectedYear}-New_Release_Report_US_`);
                setIsLoading(false);
                setIsOpen(false);
            })
            .catch(error => {
                setIsLoading(false);
                setIsOpen(false);
                addToast({
                    summary: ERROR_MESSAGE,
                    severity: 'error',
                });
            });
    };

    const getYears = () => {
        const list = [];
        for (let i = START_YEAR; i <= END_YEAR; i++) {
            list.push({value: i, label: i});
        }
        list.push({value: MOCK_YEAR, label: MOCK_YEAR});
        return list;
    };

    const getContent = () => {
        return (
            <div className="nexus-c-right-repository-release-report-content">
                <Select
                    value={{value: selectedYear, label: selectedYear}}
                    className="nexus-c-right-repository-release-report__year-dropdown"
                    onChange={val => setSelectedYear(val.value)}
                    options={getYears()}
                    classNamePrefix="nexus-c-nexus-select"
                    isSearchable
                />
                <Select
                    value={{value: selectedMonth, label: selectedMonth}}
                    className="nexus-c-right-repository-release-report__month-dropdown"
                    onChange={val => setSelectedMonth(val.value)}
                    options={MONTHS.map(m => ({value: m, label: m}))}
                    classNamePrefix="nexus-c-nexus-select"
                    isSearchable
                />
                <Button
                    isLoading={isLoading}
                    className="nexus-c-right-repository-release-report-content__btn"
                    onClick={onCreateReport}
                    appearance="primary"
                    shouldFitContainer
                >
                    {CREATE_REPORT}
                </Button>
            </div>
        );
    };
    return (
        <div className="nexus-c-right-repository-release-report">
            <Popup
                id="nexus-c-right-repository-release-report__popup"
                boundariesElement="scrollParent"
                placement="bottom-start"
                isOpen={isOpen}
                onClose={() => setIsOpen(false)}
                content={() => getContent()}
                trigger={triggerProps => (
                    <Button {...triggerProps} isSelected={isOpen} onClick={() => setIsOpen(!isOpen)} isDisabled={isDisabled} shouldFitContainer>
                        <span className="nexus-c-right-repository-release-report-button__title">{NEW_RELEASE_REPORT}</span>
                        <AtlaskitMoreIcon />
                    </Button>
                )}
            />
        </div>
    );
};

AvailsTableReleaseReport.propTypes = {
    addToast: PropTypes.func,
    activeTab: PropTypes.string,
    selectedRows: PropTypes.array.isRequired,
    totalRows: PropTypes.number.isRequired,
    prePlanRightsCount: PropTypes.number,
    planningRightsCount: PropTypes.number,
};

AvailsTableReleaseReport.defaultProps = {
    addToast: () => null,
    activeTab: RIGHTS_TAB,
    prePlanRightsCount: 0,
    planningRightsCount: 0,
};

export default withToasts(AvailsTableReleaseReport);
