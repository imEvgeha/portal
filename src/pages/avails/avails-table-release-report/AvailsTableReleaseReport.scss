.nexus-c-right-repository-release-report {
    margin-left: 10px;

    .nexus-c-right-repository-release-report-button__title {
        margin: 0 4px;
    }
}
#nexus-c-right-repository-release-report__popup {
    overflow: unset;
    .nexus-c-right-repository-release-report-content {
        width: 175px;
        padding: 20px;
        .nexus-c-right-repository-release-report__month-dropdown {
            margin: 15px 0;
        }
        .nexus-c-right-repository-release-report__month-dropdown
            > .nexus-c-nexus-select__menu
            > .nexus-c-nexus-select__menu-list {
            max-height: 300px;
            height: 300px;
        }
    }
}

.nexus-c-right-repository-release-report-content__btn {
    width: 100%;
    background: #6366F1;
    border: 1px solid #6366F1;
}