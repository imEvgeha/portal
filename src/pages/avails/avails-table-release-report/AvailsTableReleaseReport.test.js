import React from 'react';
import {shallow} from 'enzyme';
import AvailsTableReleaseReport from './AvailsTableReleaseReport';

describe('AvailsTableReleaseReport', () => {
    let wrapper = null;

    beforeEach(() => {
        wrapper = shallow(<AvailsTableReleaseReport />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
