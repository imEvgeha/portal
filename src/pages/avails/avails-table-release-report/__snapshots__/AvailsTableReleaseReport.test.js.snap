// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvailsTableReleaseReport should match snapshot 1`] = `
<div
  className="nexus-c-right-repository-release-report"
>
  <Button
    className="p-button-outlined p-button-secondary"
    disabled={false}
    icon="po po-expand-section"
    iconPos="right"
    id="nexus-c-right-repository-release-report__popup"
    label="Report"
    onClick={[Function]}
  />
  <ForwardRef
    className="report-overlay-panel"
    id="opReport"
    style={
      Object {
        "width": "auto",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="col-12 text-center"
      >
        <div
          className="nexus-c-right-repository-release-report-content"
        >
          <div
            className="row mb-1"
          >
            <div
              className="col-12"
            >
              <Dropdown
                appendTo="self"
                className="nexus-c-right-repository-release-report__year-dropdown"
                columnClass="col-12"
                filter={true}
                id="ddlYears"
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "label": 1970,
                      "value": 1970,
                    },
                    Object {
                      "label": 1971,
                      "value": 1971,
                    },
                    Object {
                      "label": 1972,
                      "value": 1972,
                    },
                    Object {
                      "label": 1973,
                      "value": 1973,
                    },
                    Object {
                      "label": 1974,
                      "value": 1974,
                    },
                    Object {
                      "label": 1975,
                      "value": 1975,
                    },
                    Object {
                      "label": 1976,
                      "value": 1976,
                    },
                    Object {
                      "label": 1977,
                      "value": 1977,
                    },
                    Object {
                      "label": 1978,
                      "value": 1978,
                    },
                    Object {
                      "label": 1979,
                      "value": 1979,
                    },
                    Object {
                      "label": 1980,
                      "value": 1980,
                    },
                    Object {
                      "label": 1981,
                      "value": 1981,
                    },
                    Object {
                      "label": 1982,
                      "value": 1982,
                    },
                    Object {
                      "label": 1983,
                      "value": 1983,
                    },
                    Object {
                      "label": 1984,
                      "value": 1984,
                    },
                    Object {
                      "label": 1985,
                      "value": 1985,
                    },
                    Object {
                      "label": 1986,
                      "value": 1986,
                    },
                    Object {
                      "label": 1987,
                      "value": 1987,
                    },
                    Object {
                      "label": 1988,
                      "value": 1988,
                    },
                    Object {
                      "label": 1989,
                      "value": 1989,
                    },
                    Object {
                      "label": 1990,
                      "value": 1990,
                    },
                    Object {
                      "label": 1991,
                      "value": 1991,
                    },
                    Object {
                      "label": 1992,
                      "value": 1992,
                    },
                    Object {
                      "label": 1993,
                      "value": 1993,
                    },
                    Object {
                      "label": 1994,
                      "value": 1994,
                    },
                    Object {
                      "label": 1995,
                      "value": 1995,
                    },
                    Object {
                      "label": 1996,
                      "value": 1996,
                    },
                    Object {
                      "label": 1997,
                      "value": 1997,
                    },
                    Object {
                      "label": 1998,
                      "value": 1998,
                    },
                    Object {
                      "label": 1999,
                      "value": 1999,
                    },
                    Object {
                      "label": 2000,
                      "value": 2000,
                    },
                    Object {
                      "label": 2001,
                      "value": 2001,
                    },
                    Object {
                      "label": 2002,
                      "value": 2002,
                    },
                    Object {
                      "label": 2003,
                      "value": 2003,
                    },
                    Object {
                      "label": 2004,
                      "value": 2004,
                    },
                    Object {
                      "label": 2005,
                      "value": 2005,
                    },
                    Object {
                      "label": 2006,
                      "value": 2006,
                    },
                    Object {
                      "label": 2007,
                      "value": 2007,
                    },
                    Object {
                      "label": 2008,
                      "value": 2008,
                    },
                    Object {
                      "label": 2009,
                      "value": 2009,
                    },
                    Object {
                      "label": 2010,
                      "value": 2010,
                    },
                    Object {
                      "label": 2011,
                      "value": 2011,
                    },
                    Object {
                      "label": 2012,
                      "value": 2012,
                    },
                    Object {
                      "label": 2013,
                      "value": 2013,
                    },
                    Object {
                      "label": 2014,
                      "value": 2014,
                    },
                    Object {
                      "label": 2015,
                      "value": 2015,
                    },
                    Object {
                      "label": 2016,
                      "value": 2016,
                    },
                    Object {
                      "label": 2017,
                      "value": 2017,
                    },
                    Object {
                      "label": 2018,
                      "value": 2018,
                    },
                    Object {
                      "label": 2019,
                      "value": 2019,
                    },
                    Object {
                      "label": 2020,
                      "value": 2020,
                    },
                    Object {
                      "label": 2021,
                      "value": 2021,
                    },
                    Object {
                      "label": 2022,
                      "value": 2022,
                    },
                    Object {
                      "label": 2023,
                      "value": 2023,
                    },
                    Object {
                      "label": 2024,
                      "value": 2024,
                    },
                    Object {
                      "label": 2025,
                      "value": 2025,
                    },
                    Object {
                      "label": 2026,
                      "value": 2026,
                    },
                    Object {
                      "label": 2027,
                      "value": 2027,
                    },
                    Object {
                      "label": 2028,
                      "value": 2028,
                    },
                    Object {
                      "label": 2029,
                      "value": 2029,
                    },
                    Object {
                      "label": 2030,
                      "value": 2030,
                    },
                    Object {
                      "label": 2099,
                      "value": 2099,
                    },
                  ]
                }
                placeholder="Select one option"
                value={2022}
              />
            </div>
          </div>
          <div
            className="row mb-1"
          >
            <div
              className="col-12"
            >
              <Dropdown
                appendTo="self"
                className="nexus-c-right-repository-release-report__month-dropdown"
                columnClass="col-12"
                filter={true}
                id="ddlMonths"
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "label": "January",
                      "value": "January",
                    },
                    Object {
                      "label": "February",
                      "value": "February",
                    },
                    Object {
                      "label": "March",
                      "value": "March",
                    },
                    Object {
                      "label": "April",
                      "value": "April",
                    },
                    Object {
                      "label": "May",
                      "value": "May",
                    },
                    Object {
                      "label": "June",
                      "value": "June",
                    },
                    Object {
                      "label": "July",
                      "value": "July",
                    },
                    Object {
                      "label": "August",
                      "value": "August",
                    },
                    Object {
                      "label": "September",
                      "value": "September",
                    },
                    Object {
                      "label": "October",
                      "value": "October",
                    },
                    Object {
                      "label": "November",
                      "value": "November",
                    },
                    Object {
                      "label": "December",
                      "value": "December",
                    },
                  ]
                }
                placeholder="Select one option"
                value="September"
              />
            </div>
          </div>
          <div
            className="row mb-1"
          >
            <div
              className="col-12"
            >
              <Button
                className="p-button-outlined"
                label="Create Report"
                loading={false}
                onClick={[Function]}
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </ForwardRef>
</div>
`;
